### PR TITLE
8307462: [REDO] VmObjectAlloc is not generated by intrinsics methods which allocate objects

### PR DESCRIPTION
**Pull request by lmesnik**
The fix adds posting VmObjectAlloc events by Unsafe.allocateInstance(Class<?> cls). The previous attempt to post event directly from 'LibraryCallKit::inline_unsafe_allocate()' cause performance regression even if jvmti event is not enabled.  Some optimizations have been disabled just because possible usage and escaping of newly allocated  object.
So event posting is doing by returning to interpreter if events are enabled.

I verified that that performance (run locally only) of
org.renaissance.jdk.streams.JmhScrabble.runOperation
doesn't change if events are not enabled.

There might be other intrinsics like 'LibraryCallKit::inline_unsafe_newArray()' where VM allocate memory. I'm going to file separate issue to find and fix them.

Many thanks to Tobias H. for proposed solution.

Testing with all tiers.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307462](https://bugs.openjdk.org/browse/JDK-8307462): [REDO] VmObjectAlloc is not generated by intrinsics methods which allocate objects (**Bug** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [64871b91](https://git.openjdk.org/jdk/pull/15110/files/64871b91de8ae9a01ad7f1ccdc613179f726adf3)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [64871b91](https://git.openjdk.org/jdk/pull/15110/files/64871b91de8ae9a01ad7f1ccdc613179f726adf3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15110/head:pull/15110` \
`$ git checkout pull/15110`

Update a local copy of the PR: \
`$ git checkout pull/15110` \
`$ git pull https://git.openjdk.org/jdk.git pull/15110/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15110`

View PR using the GUI difftool: \
`$ git pr show -t 15110`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15110.diff">https://git.openjdk.org/jdk/pull/15110.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15110#issuecomment-1663217642)

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
This looks okay to me.
It needs to be reviewed by someone from the compiler team.
Thanks,
Serguei

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -2820,6 +2828,23 @@ bool LibraryCallKit::inline_unsafe_allocate() {
...
+    Node* should_post_vm_object_alloc = make_load(this->control(), addr, TypeInt::INT, T_INT, MemNode::unordered);
+    Node* chk = _gvn.transform(new CmpINode(should_post_vm_object_alloc, intcon(0)));
+    Node* tst = _gvn.transform(new BoolNode(chk, BoolTest::eq));
+    {
+      BuildCutout unless(this, tst, PROB_MAX);
+      uncommon_trap(Deoptimization::Reason_intrinsic,
+                    Deoptimization::Action_make_not_entrant);
+    }
+    if (stopped())
+      return true;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:2845]:
```suggestion
    if (stopped()) {
      return true;
    }
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by lmesnik** [comment]
@sspitsyn, @TobiHartmann Thank you for review.
/integrate

**Pull request closed without merging**